src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -31,18 +31,36 @@ public static class ActivatorUtilities
             [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type instanceType,
             params object[] parameters)
         {
-            int bestLength = -1;
-            bool seenPreferred = false;
+            if (provider == null)
+            {
+                throw new ArgumentNullException(nameof(provider));
+            }
 
-            ConstructorMatcher bestMatcher = default;
+            if (instanceType.IsAbstract)
+            {
+                throw new InvalidOperationException(SR.CannotCreateAbstractClasses);
+            }
 
-            if (!instanceType.IsAbstract)
+            IServiceProviderIsService? serviceProviderIsService = provider.GetService<IServiceProviderIsService>();
+            // if container supports using IServiceProviderIsService, we try to find the longest ctor that
+            // (a) matches all parameters given to CreateInstance
+            // (b) matches the rest of ctor arguments as either a parameter with a default value or as a service registered
+            // if no such match is found we fallback to the same logic used by CreateFactory which would only allow creating an
+            // instance if all parameters given to CreateInstance only match with a single ctor
+            if (serviceProviderIsService != null &&
+                serviceProviderIsService.IsService(typeof(IServiceProviderIsService)))
             {
+                int bestLength = -1;
+                bool seenPreferred = false;
+
+                ConstructorMatcher bestMatcher = default;
+                bool multipleBestLengthFound = false;
+
                 foreach (ConstructorInfo? constructor in instanceType.GetConstructors())
                 {
                     var matcher = new ConstructorMatcher(constructor);
                     bool isPreferred = constructor.IsDefined(typeof(ActivatorUtilitiesConstructorAttribute), false);
-                    int length = matcher.Match(parameters);
+                    int length = matcher.Match(parameters, serviceProviderIsService!);
```

Comment:
```suggestion
                    int length = matcher.Match(parameters, serviceProviderIsService);
```

This has already been checked for `null`.